HAL interfaces for resize, warpAffine and warpPerspective

mshabunin opened the pull request:
`resize`, `warpAffine` and `warpPerspective` interfaces.

IPP blocks have been removed because there are no `ippiWarpAffine_<mode>`  and `ippiWarpPerspective_<mode>` functions in 8 and 9 versions of IPP (only in 7th), thus whole parts should be rewritten.

vpisarev commented on modules/imgproc/include/opencv2/imgproc/hal/hal.hpp:
```diff
@@ -60,6 +60,21 @@ struct  CV_EXPORTS MorphContext
 };
 
 
+CV_EXPORTS void resize(int src_type,
```
dst_width and dst_height should be added to preserve OpenCV semantics. Because of floating-point arithmetics, sometimes we can not rely on HAL (or any library) to compute the size of destination image, we compute it ourselves and pass it explicitly.

1) if dst_width and dst_height are 0's, use inv_scale_x and inv_scale_y to compute it.
2) if dst_width and dst_height are non 0's, but inv_scale_x and inv_scale_y are 0's, compute them as ratios.
3) if all the parameters are non 0's, fill out the specified destination image - some parts of the source image might be cut off or replicated; the original OpenCV code does that. There can probably be a check that inv_scale_x and dst_width/src_width, inv_scale_y and dst_height/src_height are not very different.
4) in any other case report an error.

vpisarev commented on modules/imgproc/src/imgwarp.cpp:
```diff
@@ -3226,24 +3228,7 @@ void cv::resize( InputArray _src, OutputArray _dst, Size dsize,
-
-    CV_Assert( ssize.area() > 0 );
-    CV_Assert( dsize.area() > 0 || (inv_scale_x > 0 && inv_scale_y > 0) );
-    if( dsize.area() == 0 )
```
this semantics should be kept

vpisarev commented on modules/imgproc/src/imgwarp.cpp:
```diff
@@ -3226,24 +3228,7 @@ void cv::resize( InputArray _src, OutputArray _dst, Size dsize,
         resizeArea_<double, double>, 0
     };
 
-    Size ssize = _src.size();
```
please, restore the original semantics; above it's explained why it is important

vpisarev commented on modules/imgproc/src/imgwarp.cpp:
```diff
@@ -5410,63 +5431,10 @@ class WarpAffineInvoker :
 };
 
 
-#if defined (HAVE_IPP) && IPP_VERSION_X100 >= 810 && IPP_DISABLE_BLOCK
```
@mshabunin, we do not want to be accused by Intel that we remove IPP from OpenCV. We promised no to do so. Can you please, retain the code, comment it off and supply explaining comment that the API is not available anymore and thus this code needs to be updated? The same should be done for warpPerspective and other places where IPP was used and now is not

mshabunin commented:
@vpisarev , I've updated the patch according to your comments.

Other changes include:
- HAL documentation generation support
- Added documentation for replacement interface
- Added several missing macro definitions
- Renamed `MorphContext` class
